File: App/src/main/java/com/mercury/zippit/net/ZippitChannelInitialiser.java
package com.mercury.zippit.net;

import com.mercury.zippit.net.codec.handshake.HandshakeDecoder;
import com.mercury.zippit.net.codec.handshake.HandshakeEncoder;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelPipeline;
import io.netty.channel.socket.SocketChannel;

import java.util.Objects;

/**
 * @author Harrison, Alias: Hc747, Contact: harrisoncole05@gmail.com
 * @version 1.0
 * @since 26/10/17
 */
public final class ZippitChannelInitialiser extends ChannelInitializer<SocketChannel> {

	private final ZippitHandler handler;

	public ZippitChannelInitialiser(ZippitHandler handler) {
		this.handler = Objects.requireNonNull(handler, "handler");
	}

	@Override
	protected void initChannel(SocketChannel ch) {
		ChannelPipeline pipeline = ch.pipeline();
		pipeline.addLast(HandshakeDecoder.class.getSimpleName(), new HandshakeDecoder());
		pipeline.addLast(HandshakeEncoder.class.getSimpleName(), new HandshakeEncoder());
		pipeline.addLast(ZippitHandler.class.getSimpleName(), handler);
	}
}

File: App/src/main/java/com/mercury/zippit/net/ZippitChannelInitialiser.kt
package com.mercury.zippit.net

import com.mercury.zippit.net.codec.service.ServiceDecoder
import com.mercury.zippit.net.codec.service.ServiceEncoder
import io.netty.channel.ChannelInitializer
import io.netty.channel.socket.SocketChannel
import io.netty.handler.codec.protobuf.ProtobufVarint32FrameDecoder
import io.netty.handler.codec.protobuf.ProtobufVarint32LengthFieldPrepender

/**
 * @author Harrison, Alias: Hc747, Contact: harrisoncole05@gmail.com
 * @version 1.0
 * @since 26/10/17
 */
class ZippitChannelInitialiser(private val handler: ZippitHandler) : ChannelInitializer<SocketChannel>() {

    override fun initChannel(ch: SocketChannel) {
        val pipeline = ch.pipeline()

        val decoder = ServiceDecoder()
        val encoder = ServiceEncoder()

        pipeline.addLast(ProtobufVarint32FrameDecoder::class.java.simpleName, ProtobufVarint32FrameDecoder())
        pipeline.addLast(ServiceDecoder::class.java.simpleName, decoder)

        pipeline.addLast(ProtobufVarint32LengthFieldPrepender::class.java.simpleName, ProtobufVarint32LengthFieldPrepender())
        pipeline.addLast(ServiceEncoder::class.java.simpleName, encoder)

        pipeline.addLast(ZippitHandler::class.java.simpleName, handler)
    }
}
